File: python/medium/Triangle.py
class Solution:
    def __init__(self):
        self.minimum = None
        
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        queue = {0:triangle[0][0]}
        for t in triangle[1:]:
            new_queue = {}
            for index in range(len(t)):
                new_queue[index] = []
                if index - 1 >= 0:
                    new_queue[index].append(queue[index-1]+t[index])
                if index < len(queue):
                    new_queue[index].append(queue[index]+t[index])
                new_queue[index] = min(new_queue[index])
            queue = new_queue
        return min(queue.values())

File: cpp/medium/Triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int levels = triangle.size();
        for (int level = 1; level < levels; ++level) {
            for (int idx = 0; idx < triangle[level].size(); ++idx) {
                if (idx == 0) triangle[level][idx]+=triangle[level-1][0];
                else if (idx == triangle[level].size()-1) triangle[level][idx]+=triangle[level-1][idx-1];
                else triangle[level][idx]+=min(triangle[level-1][idx-1], triangle[level-1][idx]);
            }
        }
        return *min_element(triangle[levels-1].begin(), triangle[levels-1].end());
    }

};
